### ISSUE TITLE
Move the plugin configuration to stateless

### ISSUE THREAD

**Issue by Desvelao** [wazuh/wazuh-dashboard-plugins#6201]
**Description**
As part of the issue that pretends to remove the stateful of the Wazuh dashboard plugins, we need to adapt the way the plugin's configuration is defined through the current wazuh.yml file.

The configuration will be saved in a saved object. This should support deploys with multiple instances of the Wazuh dashboard with shared or independent configuration. More info: https://github.com/wazuh/wazuh-dashboard-plugins/issues/6201#issuecomment-1862289651
| Type | ID |
| --- | --- |
| `wazuh-dashboard-plugins-config` | id of instance |

The sensitive data related to API host entries must be protected through encryption. The data encryption must use a secret. More info: https://github.com/wazuh/wazuh-dashboard-plugins/issues/6201#issuecomment-1884739017

The configuration management (edit settings, add API hosts) must be only allowed for the privileged users regarding the platform login data. More info: https://github.com/wazuh/wazuh-dashboard-plugins/issues/6201#issuecomment-1956913929

The API host entries must be managed through the UI.

Tasks:
- [x] Analyze the possibilities
- [x] Implement the configuration management:
  - [x] Create configuration service in the core plugin (backend-side)
  - [x] Create store repository (saved-objects) in the core plugin (backend-side)
  - [x] Create configuration service in the core plugin (frontend-side)
  - [x] Adapt the configuration management
    - [x] Adapt endpoint controllers
    - [x] Adapt jobs in the main plugin
    - [x] Refactor services that used the old configuration
      - [x] `manageHosts`
      - [x] `serverAPIClient`
      - [x] `cacheAPIUserAllowRunAs`
      - [x] `serverAPIHostEntries`
      - [x] `updateConfigurationFile`
  - [x] Remove management of `wazuh.yml` file
  - [x] Remove the `getConfiguration` service
  - [x] Move the services related to settings to integrate into the Configuration service
  - [x] Adapt the configuration through `App Settings`
  - [x] Enhance the Security factory of the core plugin on the backend side
  - [x] Protect the routes related to managing the configuration with the privileged permissions
  - [x] Add a mechanism to encrypt the sensitive data of API hosts configuration
  - [x] Create a security dashboard service on the frontend side to manage the permissions related to the privileged user
- [x] Nice to have. Create a utility to port the changes to the new system.

Other related issues:
- https://github.com/wazuh/wazuh-dashboard-plugins/issues/4159
- https://github.com/wazuh/wazuh-dashboard-plugins/issues/3637

**Comment by Desvelao** [comment]
# Analysis

## Save the configuration in a saved object

Pros:
- React to real-time configuration

Cons:
- Hard to share or port the configuration to another Wazuh dashboard
- Sensitive data as API server password is accessible for any user with read permissions of the saved object

## Use the configuration of the platform (opensearch_dashboards.yml)

Pros:
- Configuration is not visible. The sensitive data could be stored in the keystore (this needs research)
- Easily share the configuration

Cons:
- Require renaming the settings. The plugin configuration settings need a namespace (prefix).
- Can not react to real-time changes. Need a restart.

# Comparation

| Description | Current | Store in saved object | Platform configuration |
| --- | --- | --- | --- |
| React to real-time changes (do not need a restart) | :white_check_mark: (some settings need a restart) | :white_check_mark: | :x: |
| Share configuration easily | :white_check_mark: | :x: | :white_check_mark: |
| Protected data | :white_check_mark: | :x: | :white_check_mark: |

**Comment by Desvelao** [comment]
We had a meeting today and we decided to explore the possibility of creating a similar initiation process for the security of the Wazuh indexer to set the plugins settings.

**Comment by Desvelao** [comment]
# Approach: Store the configuration in a saved object

The intention is to store the plugin configuration in a saved object instead of the current configuration file (`wazuh.yml`). This approach causes the data to be more exposed, so we must take some actions to protect the sensitive data as the host.

## Design

I will create an interface to communicate with the configuration in the server and fronted side that will use the current plugin setting schema (extending as necessary).

Backend:
- Store in a saved object

Frontend:
- Store in memory
- Ability to retrieve new configuration

These new services will be created in the core plugin and exposed to the rest of the plugins that could register their own settings.

**Comment by Desvelao** [comment]
I was exploring the idea to manage the API host entries from the App Settings configuration, but, it seems overcomplicated due to the layout and manage the password values.

I undone this to move to the Server APIs application where the server API are listed and can be added, updated and removed.

As the ability to manage the API host from the UI is new, we need to specify the requirements for a user can execute these actions. This affects the UI buttons and the backend side that manages the API host entries.

**Comment by Desvelao** [comment]
The API host data is saved in the saved object of the plugin's configuration. This saved object can be retrieved from any user with permission about the index where this is located. As the API host data will save some sensitive data, we could need to implement a mechanism to encrypt/decrypt the data in the saved object. In the previous configuration system in the `wazuh.yml` file, the API host data is in plain text.

I recently added the ability to the ConfigurationStore service on the backend side to map/unmap the password values for the API host entries.

We need to decide the encriptation method, but one of the possibilities allows to the user to configure the method or options to this data is encrypted. We need to decide how these options are defined taking into account we want to the plugins be stateless (no required files to work). Moreover we need to take in account the process when the options to encrypt the data changes, a migration process (automatic or manual by the user) should be done to keep the previous data.

**Comment by Desvelao** [comment]
# Manage the API host entries

The API host entries can be managed through the `Server APIs` application.

The `Add new` button was replaced to `Add API host` which opens a flyout that contains a form with the required API host data.

TODO: screenshot

Moreover, a button to adding or removing an API host entry was added to the API hosts table.

TODO: screenshot

The `Add API host`, editing or removing buttons are protected to a privileged user. The API endpoints to configure the API hosts were protected with the same requirement.
The user must have the `all_access` role.

**Comment by Desvelao** [comment]
# Privilegied user

The plugin's configuration management is restricted to privileged users.

The privileged users can be defined through RBAC using the roles of the Wazuh dashboard.

A new plugin setting was added to configure the platform user roles that are privileged for the Wazuh plugins.

| Plugin setting | Type | Description | Default value |
| --- | --- | --- | --- |
| `wazuh_core.security.administrator.roles` | string | Define the platform roles of the privilegied user| `["all_access"]` | 

Example:
```yml
wazuh_core.security.administrator.roles: ["all_access", "wazuh_admin"]
```

A new endpoint was created that returns if the current user is a privileged user or not.

```
GET /utils/account/is-admin
```

Responses:
- User is admin: 
  - HTTP code: `200`
```json
{
is_admin: true
}
```
- User is `NOT` admin: 
  - HTTP code: `200`
```json
{
is_admin: false,
message: "<Permission denied reason>"
}
```

**Comment by Desvelao** [comment]
# Migrate or pre-setup the configuration

I am researching how to migrate or pre-setup the configuration using the previous configuration through the `wazuh.yml` configuration file to the new system stored in a saved object.

## Migration task on Wazuh dashboard initiation

I added a mechanism to update the configuration using the configuration file located at `data/wazuh/config/wazuh.yml` (location of configuration in the previous plugins).

This mechanism reads the configuration file and saves it to the new system. Then the file is renamed to `wazuh.yml.backup` to avoid the following restart triggering the migration process.

This process is indicated to migrate the configuration.

Due to the renaming of the file, if the configuration file was mounted a volume in the Wazuh dashboard installions through Docker could cause an error.

## Script

I added a script to update the configuration reading the `wazuh.yml` configuration file. This script uses the API endpoints related to the plugin configuration to set the settings on the configuration file.

The script is written in script, but the idea is to expose a `sh` wrapper.

This utility could help to setup (or migrate) some plugin configuration defined in a `wazuh.yml` configuration file.

This process needs to use a privilegied user.

**Comment by Desvelao** [comment]
# Sync meeting

## Show

- Configuration encryption.
  - New plugin setting
- Administrator user roles

## Discussion

- No multiple instances of Wazuh dashboard (use case: worker fetch data for statistics and monitoring)

- Replace the current protection of privileged user (`administrator` API role) for the platform role:
  - Sample data: view UI, add and remove
  - App Settings
  - Delete PDF report
  
- Protect view actions related to privileged user:
  - App Settings
  - Deploy new agent option to set the enrollment.dns
  - Server APIs?
  
- Migration process
  - Server side
  - Script
    - Where should it be included the binary: plugin or platform?

  > Changes could require to a restart of a Wazuh dashboard

## Action items
- Administrator role: research about cluster permissions instead of roles
- Ignore the migration process when the saved object has some settings defined
- See parallel monitoring and migration
- Support multiple instances of Wazuh dashboards
- Disable button user has no permissions: Deploy new agent option to set the `enrollment.dns`
- Where should it be included in the binary: plugin or platform? Plugin. Create an issue to move to `<WAZUH_DASHBOARD>/bin`

**Comment by Desvelao** [comment]
# Research

One of the action items of the last meeting was:
> Administrator role: research about cluster permissions instead of roles
whose intention is to consider the administrator user based on the cluster permissions.

Currently, the security data of the user is getting through `GET /_opendistro/_security/api/account` but it is does not include data about the cluster permissions. But it includes the roles of the user.

The role information can be used with the security plugin API to get the cluster permissions using the endpoint:
```
GET /api/v1/configuration/roles/{id}
```
whose response https://github.com/opensearch-project/security-dashboards-plugin/blob/v1.13.0.0/server/routes/index.ts#L273-L290
```
   * for role, response will look like:
   * {
   *   "reserved": true,
   *   "hidden": false,
   *   "description": "Allow full access to all indices and all cluster APIs",
   *   "cluster_permissions": ["*"],
   *   "index_permissions": [{
   *     "index_patterns": ["*"],
   *     "fls": [],
   *     "masked_fields": [],
   *     "allowed_actions": ["*"]
   *   }],
   *   "tenant_permissions": [{
   *     "tenant_patterns": ["*"],
   *     "allowed_actions": ["kibana_all_write"]
   *   }],
   *   "static": false
   * }
```
The `cluster_permissions` can include an action group or specific cluster permissions. We could get the specific cluster permissions of an action group with the API request:
```
GET /api/v1/configuration/actiongroups/{id}
```

In summary, to check if an user could have the desired cluster permission we could need to do multiple API requests:
- Get user info
- Get role info (for each role)
- Get an action group (for each group)
> We could use the generic API endpoints to get data about the roles or actions ( if all the data is returned). This could reduce the API request to 3.

**Comment by Desvelao** [comment]
# Changes

- Renamed platform setting: `wazuh_core.configuration.encryption_password` to `wazuh_core.configuration.encryption_key`

**Comment by Desvelao** [comment]
# Meeting

team: documenation

I asked to the documentation team a meeting to define the changes to apply in the documentation and take into account some considerations to organize the content.

## Topics
- Organize the new configuration system
  - Removed `wazuh.yml`
  - Use `wazuh.yml` to load configuration
  - Platform settings
  - API dashboard
  - References to the configuration file: what to do?

**Comment by Desvelao** [comment]
# Research - Administrator: approach to display the Security plugin

We could use to consider the administrator user the same approach that is used to display the Security plugin that is only visible for users with permissions to access to the API.

The Security plugin of OpenSearch Dashboards does a request to check the permissions and register the application depending on the value of `has_api_access` field.

Endpoint: `GET /api/v1/restapinfo`
Register Security application depending on the `has_api_access` value: https://github.com/opensearch-project/security-dashboards-plugin/blob/2.11.0.0/public/plugin.ts#L99-L100
Request definition and check response value: https://github.com/opensearch-project/security-dashboards-plugin/blob/2.11.0.0/public/plugin.ts#L54C16-L63
Request: https://github.com/opensearch-project/security-dashboards-plugin/blob/2.11.0.0/public/plugin.ts#L89

Following this approach, we could create a function with the similar check and use it to consider the administrator user.

```ts
async function isAdministratorUser(
  context: RequestHandlerContext,
  request: OpenSearchDashboardsRequest,
) {
  const response = await context.security_plugin.esClient
    .asScoped(request)
    .callAsCurrentUser('opensearch_security.restapiinfo');

  if (!response.has_api_access) {
    throw new Error(`User has no permission.`);
  }
}
```

This approach removes the requirements to implement our own system based on roles or cluster permissions.

If we decide to continue with this approach, then we need to clarify the permissions requirements that cause the `has_api_access` returns `true`.

**Comment by Desvelao** [comment]
# Changes

- Add a cache to the ConfigurationStore service of the backend side that avoids the repetition of requests to get the content of the saved object.

**Comment by Desvelao** [comment]
# Changes
- Minor fixes
- Adapt the monitoring and statistics jobs
- Generate suite of manual tests
- Update development branch with the last changes
- Resolve some TODOs
- Fix unit tests
- Rename references to API host by API connection in the Server APIs application and ManageHosts service

**Comment by Desvelao** [comment]
# Changes

- Move some AngularJS dependencies to the component in the ApiTable component
- Create API endpoint to create API connections:
    - POST /hosts/apis/{id}
- Adapt the API request done by the form of API connection
- Adapt setup-configuration script to use the new API endpoint
- Add method to create API connection to the ManageHosts service

**Comment by Desvelao** [comment]
# Changes
- Add some tests related to `wazuh-utils`
- Rework the setup of configuration based on importing the file through a new API endpoint. This does unnecessary the addition to own script.

**Comment by Desvelao** [comment]
# Changes

- Renamed the type of configuration saved object from `plugins-configuration` to `wazuh-dashboard-plugins-config`
